### PR TITLE
 Toggles for special MobSpawner classes per world

### PR DESCRIPTION
**Pull request by jpenilla**
Paper-config-files.patch:
- Adds a getBoolean with a Predicate to PaperWorldConfig to allow for the default behavior to respect level type in the same way as vanilla

Allow-toggling-special-MobSpawners-per-world.patch:
- Adds toggles for each MobSpawner subclass instead of using an ImmutableList of all of them for dim type overworld and an empty ImmutableList for any other.

Make-Wandering-Traders-spawn-below-the-roof-if-confi.patch:
- Adds a fix so that when Wandering Traders spawn in the Nether, they will spawn below the roof like other nether mobs instead of on top of it

### CODE REVIEW

File: Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+diff --git a/src/main/java/net/minecraft/server/MobSpawnerTrader.java b/src/main/java/net/minecraft/server/MobSpawnerTrader.java
+index 502cb0ea40b5df1962c5a6d4b57d34de9d413e7f..1956d7d65606573ff9811c6a4066403418ef57eb 100644
+--- a/src/main/java/net/minecraft/server/MobSpawnerTrader.java
++++ b/src/main/java/net/minecraft/server/MobSpawnerTrader.java
+@@ -122,11 +122,21 @@ public class MobSpawnerTrader implements MobSpawner {
+             int k = blockposition.getX() + this.a.nextInt(i * 2) - i;
+             int l = blockposition.getZ() + this.a.nextInt(i * 2) - i;
+             int i1 = iworldreader.a(HeightMap.Type.WORLD_SURFACE, k, l);
+-            BlockPosition blockposition2 = new BlockPosition(k, i1, l);
++            // Paper start - allow traders to spawn below the nether roof
```
Comment (Proximyst) [inline Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch]: Split off into own patch/PR.

File: Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
+index efe10b7f587810749038b67db14af0bf8b9b4374..d0b40afc9e503be352afc925a664b2207a480f7f 100644
+--- a/src/main/java/net/minecraft/server/WorldServer.java
++++ b/src/main/java/net/minecraft/server/WorldServer.java
+@@ -227,7 +227,24 @@ public class WorldServer extends World implements GeneratorAccessSeed {
+         this.L = new ObjectLinkedOpenHashSet();
+         this.Q = flag1;
+         this.server = minecraftserver;
+-        this.mobSpawners = list;
++        // Paper start - enable/disable MobSpawners per world
```
Comment (Proximyst) [inline Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch]: This feels like plugin territory...

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch (hunk 2 above)
Comment (jpenilla) [reply Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch]: Seems like it would need some hacky reflection to accomplish this with a plugin, given the list passed into the constructor is immutable

Comment (Proximyst) [reply Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch]: Unless there's something I'm missing, this can be done with API only: https://papermc.io/javadocs/paper/1.16/com/destroystokyo/paper/event/entity/PreSpawnerSpawnEvent.html

Comment (jpenilla) [reply Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch]: Ah, I believe that API is for AbstractMobSpawner / Mob Spawner Blocks, not the classes implementing MobSpawner, which are: MobSpawnerCat, MobSpawnerPatrol, MobSpawnerPhantom, MobSpawnerTrader, and MobSpawnerSiege

Comment (Proximyst) [reply Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch]: Bless md's awful mappings then! Feel free to resolve this as I can't from the GH app I believe

File: Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+         init();
+@@ -47,6 +52,12 @@ public class PaperWorldConfig {
+         return config.getBoolean("world-settings." + worldName + "." + path, config.getBoolean("world-settings.default." + path));
+     }
+ 
++    private boolean getBoolean(String path, Predicate<Boolean> predicate) {
++        String val = getString(path, "default").toLowerCase();
++        Boolean bool = BooleanUtils.toBooleanObject(val, "true", "false", "default");
++        return predicate.test(bool);
++    }
```
Comment (Proximyst) [inline Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch]: Move to the paper config patch.

File: Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+     private final String worldName;
++    private final Environment environment;
+     private final SpigotWorldConfig spigotConfig;
+     private final YamlConfiguration config;
+     private boolean verbose;
+ 
+-    public PaperWorldConfig(String worldName, SpigotWorldConfig spigotConfig) {
++    public PaperWorldConfig(String worldName, SpigotWorldConfig spigotConfig, Environment environment) {
+         this.worldName = worldName;
++        this.environment = environment;
```
Comment (Proximyst) [inline Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch]: Move to Paper config patch.

File: Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 287db5203f3a4995c2b09bfd9fe17734538e7466..87ad87ef41064d7ec839aab850ac0dedb9eacf0c 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -123,7 +123,7 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
+ 
+     protected World(WorldDataMutable worlddatamutable, ResourceKey<World> resourcekey, final DimensionManager dimensionmanager, Supplier<GameProfilerFiller> supplier, boolean flag, boolean flag1, long i, org.bukkit.generator.ChunkGenerator gen, org.bukkit.World.Environment env, java.util.concurrent.Executor executor) { // Paper
+         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((WorldDataServer) worlddatamutable).getName()); // Spigot
+-        this.paperConfig = new com.destroystokyo.paper.PaperWorldConfig((((WorldDataServer)worlddatamutable).getName()), this.spigotConfig); // Paper
++        this.paperConfig = new com.destroystokyo.paper.PaperWorldConfig((((WorldDataServer) worlddatamutable).getName()), this.spigotConfig, env); // Paper
```
Comment (Proximyst) [inline Spigot-Server-Patches/0564-Allow-toggling-special-MobSpawners-per-world.patch]: Move to Paper config patch.

File: Spigot-Server-Patches/0565-Make-Wandering-Traders-spawn-below-the-roof-if-confi.patch [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
++                } while (!iworldreader.getType(blockposition2).isAir());
++                do {
++                    blockposition2.c(EnumDirection.DOWN);
++                } while (iworldreader.getType(blockposition2).isAir() && blockposition2.getY() > 0);
++            }
+ 
+-            if (SpawnerCreature.a(EntityPositionTypes.Surface.ON_GROUND, iworldreader, blockposition2, EntityTypes.WANDERING_TRADER)) {
+-                blockposition1 = blockposition2;
++            if (SpawnerCreature.a(EntityPositionTypes.Surface.ON_GROUND, iworldreader, blockposition2.immutableCopy(), EntityTypes.WANDERING_TRADER)) {
++                blockposition1 = blockposition2.immutableCopy();
```
Comment (Proximyst) [inline Spigot-Server-Patches/0565-Make-Wandering-Traders-spawn-below-the-roof-if-confi.patch]: Do we need to do an immutable copy here? If so, at least put it into its own variable to avoid doing two of them.

**Review by Proximyst — CHANGES_REQUESTED** [review]

**Comment by jpenilla** [comment]
Just rebased this PR against master, the reason it's so noisy is that it touches PaperWorldConfig so almost everything else touching it needed to be updated. Besides that, this patch is a pretty simple one, and has been tested to work fine for a couple months now :)

File: Spigot-Server-Patches/0600-Allow-toggling-special-MobSpawners-per-world.patch [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
+index d0ffe5790e98b65cd6035ad529e7da8172de9029..9f90cf4c9da8b870eb6cbc468356d5a82035b2c6 100644
+--- a/src/main/java/net/minecraft/server/MinecraftServer.java
++++ b/src/main/java/net/minecraft/server/MinecraftServer.java
+@@ -402,7 +402,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
+             boolean flag = generatorsettings.isDebugWorld();
+             long i = generatorsettings.getSeed();
+             long j = BiomeManager.a(i);
+-            List<MobSpawner> list = ImmutableList.of(new MobSpawnerPhantom(), new MobSpawnerPatrol(), new MobSpawnerCat(), new VillageSiege(), new MobSpawnerTrader(iworlddataserver));
++            List<MobSpawner> list = ImmutableList.of(new MobSpawnerPhantom(), new MobSpawnerPatrol(), new MobSpawnerCat(), new VillageSiege(), new MobSpawnerTrader(iworlddataserver)); // Paper - This list is unused as we make these configurable in `Allow toggling special MobSpawners per world`. If this list changes, that patch likely needs updating.
```
Comment (Proximyst) [inline Spigot-Server-Patches/0600-Allow-toggling-special-MobSpawners-per-world.patch]:
```suggestion
+            // List<MobSpawner> list = ImmutableList.of(new MobSpawnerPhantom(), new MobSpawnerPatrol(), new MobSpawnerCat(), new VillageSiege(), new MobSpawnerTrader(iworlddataserver)); // Paper - This list is unused as we make these configurable in `Allow toggling special MobSpawners per world`. If this list changes, that patch likely needs updating.
```

File: Spigot-Server-Patches/0600-Allow-toggling-special-MobSpawners-per-world.patch [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+index b1968103165475dd4021570e3a24deb3dc3ac796..a457696b3376348a778d3507d0dbe6514d0d94e8 100644
+--- a/src/main/java/net/minecraft/server/WorldServer.java
++++ b/src/main/java/net/minecraft/server/WorldServer.java
+@@ -231,7 +231,24 @@ public class WorldServer extends World implements GeneratorAccessSeed {
+         this.L = new ObjectLinkedOpenHashSet();
+         this.Q = flag1;
+         this.server = minecraftserver;
+-        this.mobSpawners = list;
++        // Paper start - enable/disable MobSpawners per world
++        this.mobSpawners = new java.util.ArrayList<>();
```
Comment (Proximyst) [inline Spigot-Server-Patches/0600-Allow-toggling-special-MobSpawners-per-world.patch]: What if we rather use an `ImmutableList.Builder` here as Mojang used an `ImmutableList` for the original variable?

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0600-Allow-toggling-special-MobSpawners-per-world.patch (hunk 7 above)
Comment (jpenilla) [reply Spigot-Server-Patches/0600-Allow-toggling-special-MobSpawners-per-world.patch]: The list is "unused" but it is still passed to the WorldServer constructor. We could pass an empty list to the constructor, or remove the parameter, but either way increases the diff more than is needed imo.

File: Spigot-Server-Patches/0600-Allow-toggling-special-MobSpawners-per-world.patch (hunk 8 above)
Comment (jpenilla) [reply Spigot-Server-Patches/0600-Allow-toggling-special-MobSpawners-per-world.patch]: Ok
